src/core/expr/fexpr_cut.cc

```diff
@@ -110,17 +134,120 @@ class FExpr_Cut : public FExpr_Func {
         }
       }
 
-      // Cut workframe in-place
+      // Bin columns in-place
       for (size_t i = 0; i < ncols; ++i) {
-        Column coli = wf.retrieve_column(i);
-        coli = Column(Cut_ColumnImpl::make(
-                 std::move(coli), i, nbins[i], right_closed_
-               ));
-        wf.replace_column(i, std::move(coli));
+        Column col = wf.retrieve_column(i);
+
+        if (col.ltype() == LType::MU) {
+          col = Column(new ConstNa_ColumnImpl(col.nrows(), dt::SType::INT32));
+        } else {
+          if (!ltype_is_numeric(col.ltype())) {
+            throw TypeError() << "cut() can only be applied to numeric or void "
+              << "columns, instead column `" << i << "` has an stype: `"
+              << col.stype() << "`";
+          }
+
+          col = Column(CutNbins_ColumnImpl::make(
+                  std::move(col), nbins[i], right_closed_
+                ));
+        }
+        wf.replace_column(i, std::move(col));
       }
+    }
 
-      return wf;
+
+    // Bin data based on the provided interval edges.
+    void cut_bins(Workframe& wf) const {
+      const size_t ncols = wf.ncols();
+
+      if (!py_bins_.is_list_or_tuple()) {
+        throw TypeError() << "`bins` parameter must be a list or a tuple, "
+          << "instead got `" << py_bins_.typeobj() << "`";
+      }
+
+      py::oiter py_bins = py_bins_.to_oiter();
+
+      if (py_bins.size() != ncols) {
+        throw ValueError() << "Number of elements in `bins` must be equal to "
+          << "the number of columns in the frame/expression, i.e. `"
+          << ncols << "`, instead got: `" << py_bins.size() << "`";
+      }
+
+      size_t i = 0;
+      for (auto py_bin : py_bins) {
+        DataTable* dt = py_bin.to_datatable();
+        if (dt->ncols() != 1) {
+          throw ValueError() << "To bin a column `cut()` needs exactly one column "
+            << "with the bin edges, instead for the frame `" << i << "` got: "
+            << dt->ncols() << "`";
+        }
+
+        if (dt->nrows() < 2) {
+          throw ValueError() << "To bin data at least two edges are required, "
+            << "instead for the frame `" << i << "` got: `" << dt->nrows() << "`";
+        }
+
+        // Retrieve bins
+        Column bins = dt->get_column(0);
+        if (!ltype_is_numeric(bins.ltype())) {
+          throw TypeError() << "Bin edges must be provided as the numeric columns only, "
+            << "instead for the frame `" << i << "` the column stype is `"
+            << bins.stype() << "`";
+        }
+        bins.cast_inplace(SType::FLOAT64);
+        validate_bins(bins, i);
+        bins.materialize();
```

Comment:
Validation already computes all elements of the `bins` column, why not combine it with materialization?

Reply:
What do you mean by combine? Do materialization manually along with the validation? It seems to me that materialization may not be necessary if the validation fails.

Reply:
Validation here involves checking that all values are valid and in increasing order. It involves computing every element, and then comparing it with the previous element. So you could just save those values into a `std::vector<double>` at the same time.

Reply:
Ah, yes, it makes sense. Thought you want to combine it with the standard `.materialize()` method.